### ISSUE TITLE
Docker container for server crashes during start

### ISSUE THREAD

**Issue by zirpins** [di-unipi-socc/microFreshener#9]
Example output:

```
$ docker-compose up
WARNING: Some networks were defined but are not used by any service: internal
Creating microfreshener-client ... done
Creating microfreshener-server ... done
Attaching to microfreshener-server, microfreshener-client
microfreshener-server | PROD SERVER
microfreshener-server | PROD SERVER
microfreshener-server | Watching for file changes with StatReloader
microfreshener-server | Performing system checks...
microfreshener-server | 
microfreshener-server | Generate new instance
microfreshener-server | Exception in thread django-main-thread:
microfreshener-server | Traceback (most recent call last):
microfreshener-server |   File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
microfreshener-server |     self.run()
microfreshener-server |   File "/usr/local/lib/python3.8/threading.py", line 870, in run
microfreshener-server |     self._target(*self._args, **self._kwargs)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/utils/autoreload.py", line 54, in wrapper
microfreshener-server |     fn(*args, **kwargs)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/core/management/commands/runserver.py", line 117, in inner_run
microfreshener-server |     self.check(display_num_errors=True)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 387, in check
microfreshener-server |     all_issues = self._run_checks(
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 377, in _run_checks
microfreshener-server |     return checks.run_checks(**kwargs)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/core/checks/registry.py", line 72, in run_checks
microfreshener-server |     new_errors = check(app_configs=app_configs)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/core/checks/urls.py", line 13, in check_url_config
microfreshener-server |     return check_resolver(resolver)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/core/checks/urls.py", line 23, in check_resolver
microfreshener-server |     return check_method()
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/urls/resolvers.py", line 398, in check
microfreshener-server |     for pattern in self.url_patterns:
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/utils/functional.py", line 80, in __get__
microfreshener-server |     res = instance.__dict__[self.name] = self.func(instance)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/urls/resolvers.py", line 579, in url_patterns
microfreshener-server |     patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/utils/functional.py", line 80, in __get__
microfreshener-server |     res = instance.__dict__[self.name] = self.func(instance)
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/urls/resolvers.py", line 572, in urlconf_module
microfreshener-server |     return import_module(self.urlconf_name)
microfreshener-server |   File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
microfreshener-server |     return _bootstrap._gcd_import(name[level:], package, level)
microfreshener-server |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
microfreshener-server |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
microfreshener-server |   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
microfreshener-server |   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
microfreshener-server |   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
microfreshener-server |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
microfreshener-server |   File "/code/microfreshener-server/urls.py", line 30, in <module>
microfreshener-server |     path('api/', include('api.urls')),
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/django/urls/conf.py", line 34, in include
microfreshener-server |     urlconf_module = import_module(urlconf_module)
microfreshener-server |   File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
microfreshener-server |     return _bootstrap._gcd_import(name[level:], package, level)
microfreshener-server |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
microfreshener-server |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
microfreshener-server |   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
microfreshener-server |   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
microfreshener-server |   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
microfreshener-server |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
microfreshener-server |   File "/code/api/urls.py", line 6, in <module>
microfreshener-server |     from api import view
microfreshener-server |   File "/code/api/view.py", line 24, in <module>
microfreshener-server |     from microfreshener.core.refiner import KubernetesRefiner, IstioRefiner
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/microfreshener/core/refiner/__init__.py", line 2, in <module>
microfreshener-server |     from .kubernetesrefiner import KubernetesRefiner
microfreshener-server |   File "/usr/local/lib/python3.8/site-packages/microfreshener/core/refiner/kubernetesrefiner.py", line 9, in <module>
microfreshener-server |     from nested_lookup import nested_lookup
microfreshener-server | ModuleNotFoundError: No module named 'nested_lookup'
```

Seems to relate to microFreshener-core.

**Comment by Mastercuber** [comment]
The `nested_lookup` module is missing. Add this dependency to the requirements.txt file in the server directory. Now the server should launch.. But when looking into the Dockerfile in the server dir, it's not executing the migration. So this must be done manually.

**Comment by antonio-brogi** [comment]
Thanks Christian for your message

I am asking to Davide and Jacopo to check this asap.

Best

Antonio


Da: Christian Zirpins <notifications@github.com>
Inviato: lunedì 4 novembre 2019 16:45
A: di-unipi-socc/microFreshener <microFreshener@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Oggetto: [di-unipi-socc/microFreshener] Docker container for server crashes during start (#9)


Example output:

'''
$ docker-compose up
WARNING: Some networks were defined but are not used by any service: internal
Creating microfreshener-client ... done
Creating microfreshener-server ... done
Attaching to microfreshener-server, microfreshener-client
microfreshener-server | PROD SERVER
microfreshener-server | PROD SERVER
microfreshener-server | Watching for file changes with StatReloader
microfreshener-server | Performing system checks...
microfreshener-server |
microfreshener-server | Generate new instance
microfreshener-server | Exception in thread django-main-thread:
microfreshener-server | Traceback (most recent call last):
microfreshener-server | File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
microfreshener-server | self.run()
microfreshener-server | File "/usr/local/lib/python3.8/threading.py", line 870, in run
microfreshener-server | self._target(*self._args, **self._kwargs)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/utils/autoreload.py", line 54, in wrapper
microfreshener-server | fn(*args, **kwargs)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/core/management/commands/runserver.py", line 117, in inner_run
microfreshener-server | self.check(display_num_errors=True)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 387, in check
microfreshener-server | all_issues = self._run_checks(
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 377, in _run_checks
microfreshener-server | return checks.run_checks(**kwargs)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/core/checks/registry.py", line 72, in run_checks
microfreshener-server | new_errors = check(app_configs=app_configs)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/core/checks/urls.py", line 13, in check_url_config
microfreshener-server | return check_resolver(resolver)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/core/checks/urls.py", line 23, in check_resolver
microfreshener-server | return check_method()
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/urls/resolvers.py", line 398, in check
microfreshener-server | for pattern in self.url_patterns:
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/utils/functional.py", line 80, in get
microfreshener-server | res = instance.dict[self.name] = self.func(instance)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/urls/resolvers.py", line 579, in url_patterns
microfreshener-server | patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/utils/functional.py", line 80, in get
microfreshener-server | res = instance.dict[self.name] = self.func(instance)
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/urls/resolvers.py", line 572, in urlconf_module
microfreshener-server | return import_module(self.urlconf_name)
microfreshener-server | File "/usr/local/lib/python3.8/importlib/init.py", line 127, in import_module
microfreshener-server | return _bootstrap._gcd_import(name[level:], package, level)
microfreshener-server | File "", line 1014, in _gcd_import
microfreshener-server | File "", line 991, in _find_and_load
microfreshener-server | File "", line 975, in _find_and_load_unlocked
microfreshener-server | File "", line 671, in _load_unlocked
microfreshener-server | File "", line 783, in exec_module
microfreshener-server | File "", line 219, in _call_with_frames_removed
microfreshener-server | File "/code/microfreshener-server/urls.py", line 30, in
microfreshener-server | path('api/', include('api.urls')),
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/django/urls/conf.py", line 34, in include
microfreshener-server | urlconf_module = import_module(urlconf_module)
microfreshener-server | File "/usr/local/lib/python3.8/importlib/init.py", line 127, in import_module
microfreshener-server | return _bootstrap._gcd_import(name[level:], package, level)
microfreshener-server | File "", line 1014, in _gcd_import
microfreshener-server | File "", line 991, in _find_and_load
microfreshener-server | File "", line 975, in _find_and_load_unlocked
microfreshener-server | File "", line 671, in _load_unlocked
microfreshener-server | File "", line 783, in exec_module
microfreshener-server | File "", line 219, in _call_with_frames_removed
microfreshener-server | File "/code/api/urls.py", line 6, in
microfreshener-server | from api import view
microfreshener-server | File "/code/api/view.py", line 24, in
microfreshener-server | from microfreshener.core.refiner import KubernetesRefiner, IstioRefiner
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/microfreshener/core/refiner/init.py", line 2, in
microfreshener-server | from .kubernetesrefiner import KubernetesRefiner
microfreshener-server | File "/usr/local/lib/python3.8/site-packages/microfreshener/core/refiner/kubernetesrefiner.py", line 9, in
microfreshener-server | from nested_lookup import nested_lookup
microfreshener-server | ModuleNotFoundError: No module named 'nested_lookup'
'''

Seems to relate to microFreshener-core.

**Comment by dido18** [comment]
Thank you @zirpins @Mastercuber  for having opened the issues.

The issue should be resolved in commit 4342843

I have also updated the README for running the application with Docker compose.

**Comment by zirpins** [comment]
Thanks @dido18, works nicely! One Suggestion: do you really need to mount the server container /code folder as a volume in the production config? It prevents e.g. migrations done in the Docker file for the image to come into effect.
